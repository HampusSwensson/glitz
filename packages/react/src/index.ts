export { default as GlitzProvider } from './components/GlitzProvider';
export * from './styled';
