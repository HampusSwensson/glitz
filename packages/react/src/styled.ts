import { createStyled } from './styled/custom';
import { assignPredefined } from './styled/predefined';

export type { StyledComponent, StyledComponentWithRef } from './styled/create';
export type { StyledDecorator } from './styled/decorator';
export type { StyledElement, StyledElementProps } from './styled/predefined';
export type { StyledElementLike } from './styled/apply-class-name';
export type { StyledProps, StyledForwardStyle } from './styled/forward-style';
export { default as useStyle } from './styled/use-glitz';
export { default as useTheme } from './styled/use-theme';
export { applyClassName } from './styled/apply-class-name';
export { forwardStyle } from './styled/forward-style';

export const styled = assignPredefined(createStyled);
