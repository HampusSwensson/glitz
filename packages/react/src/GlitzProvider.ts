import GlitzClient from '@glitz/core';
import GlitzServer from '@glitz/core/server';
import * as React from 'react';

export type ProviderProps = {
  glitz: GlitzClient | GlitzServer;
};

export type Context = {
  glitz: GlitzClient | GlitzServer;
};

export default class Provider extends React.Component<ProviderProps> {
  public static childContextTypes = {
    glitz: () => null, //  Just pass the damn thing
  };
  private childContext: Context;
  constructor(props: ProviderProps, context: Context) {
    super(props, context);
    this.childContext = { glitz: props.glitz };
  }
  public getChildContext() {
    return this.childContext;
  }
  public render() {
    return this.props.children;
  }
}
