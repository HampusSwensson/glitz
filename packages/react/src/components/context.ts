import { createContext } from 'react';
import { GlitzClient, GlitzServer } from '@glitz/core';
import { Style, Theme } from '@glitz/type';

export const GlitzContext = createContext<GlitzClient | GlitzServer | undefined>(void 0);

export const ThemeContext = createContext<Theme | undefined>(void 0);

export type ElementPreStyle = { [tag: string]: Style[] };
export const StyleContext = createContext<ElementPreStyle | undefined>(void 0);
