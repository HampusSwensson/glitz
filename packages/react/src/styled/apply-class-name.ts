import { FunctionComponent, ComponentClass, ClassType, Component, ComponentState, ComponentType } from 'react';
import { SECRET_TYPE } from './constants';
import { WithRefProp } from './create';
import { isType, StyledType, Type } from './predefined';
import { StyledElementProps } from './types';

export interface StyledElementLike<TComponent> extends StyledType {
  [SECRET_TYPE]: Type.ElementLike;
  value: TComponent;
}

export function applyClassName<TProps extends StyledElementProps = StyledElementProps>(
  component: FunctionComponent<TProps>,
): StyledElementLike<FunctionComponent<TProps>>;

export function applyClassName<TProps extends StyledElementProps, TComponent extends ComponentClass<TProps>>(
  component: ClassType<TProps, Component<TProps, ComponentState>, TComponent>,
): StyledElementLike<TComponent>;

export function applyClassName<TProps extends StyledElementProps>(
  component: ComponentType<TProps>,
): StyledElementLike<ComponentType<TProps>> {
  return {
    [SECRET_TYPE]: Type.ElementLike,
    value: component,
  };
}

export function isElementLikeType<TProps extends StyledElementProps, TInstance>(
  type: any,
): type is StyledElementLike<ComponentType<WithRefProp<TProps, TInstance>>> {
  return isType(type) && type[SECRET_TYPE] === Type.ElementLike;
}
