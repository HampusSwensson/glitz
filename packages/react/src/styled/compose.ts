import { createContext, useContext, createElement, ReactNode, useMemo, ReactElement } from 'react';
import { DirtyStyle } from './use-glitz';

const ComposeContext = createContext<DirtyStyle[] | undefined>(void 0);

const empty = { value: void 0 };

export function useAbsorb(factory: (dynamics: DirtyStyle[] | undefined) => ReactElement) {
  const dynamics = useContext(ComposeContext);
  return createElement(ComposeContext.Provider, empty, factory(dynamics));
}

export function useForward(statics: DirtyStyle, dynamic: DirtyStyle, children: ReactNode): ReactElement {
  const dynamics = useContext(ComposeContext);
  const context = useMemo(() => ({ value: [statics, ...(dynamics ?? []), dynamic] }), [statics, dynamics, dynamic]);
  return createElement(ComposeContext.Provider, context, children);
}
