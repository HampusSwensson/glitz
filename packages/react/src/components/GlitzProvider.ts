import { GlitzClient, GlitzServer } from '@glitz/core';
import { PropsWithChildren, createElement } from 'react';
import { GlitzContext } from './context';

export type PropType = {
  glitz: GlitzClient | GlitzServer;
};

export function GlitzProvider(props: PropsWithChildren<PropType>) {
  return createElement(GlitzContext.Provider, { value: props.glitz }, props.children);
}
