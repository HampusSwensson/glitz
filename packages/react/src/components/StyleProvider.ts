import { Style } from '@glitz/type';
import { useContext, createElement, PropsWithChildren } from 'react';
import { StyledElementFunctions } from '../styled/types';
import { ElementPreStyle, StyleContext } from './context';

type PreStyleKeys = keyof StyledElementFunctions | 'universal';

export type PreStyleProps = { [tag in PreStyleKeys]?: Style };

type StyleProviderProps = PreStyleProps & { include?: PreStyleProps };

export function StyleProvider({ children, include, ...restProps }: PropsWithChildren<StyleProviderProps>) {
  const pre: ElementPreStyle = useContext(StyleContext) ?? {};

  for (const map of [include, restProps]) {
    if (map) {
      let tag: PreStyleKeys;
      for (tag in map) {
        pre[tag] = tag in pre ? [...pre[tag], map[tag]!] : [map[tag]!];
      }
    }
  }

  return createElement(StyleContext.Provider, { value: pre }, children);
}
