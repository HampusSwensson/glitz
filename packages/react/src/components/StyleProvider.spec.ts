import { GlitzClient, GlitzServer } from '@glitz/core';
import { mount } from 'enzyme';
import * as React from 'react';
import { styled } from '../styled';
import { StyleContext } from './context';
import { GlitzProvider } from './GlitzProvider';
import { StyleProvider } from './StyleProvider';

describe('StyleProvider', () => {
  it('provides style', () => {
    const Spy: React.FunctionComponent = () => {
      const providedStyle = React.useContext(StyleContext);
      expect(providedStyle).toEqual({
        universal: [{ color: 'red' }],
        a: [{ color: 'green' }, { color: 'blue' }],
        p: [{ color: 'black' }],
      });
      return null;
    };
    mount(
      React.createElement(
        GlitzProvider,
        { glitz: new GlitzClient() },
        React.createElement(
          StyleProvider,
          {
            universal: { color: 'red' },
            include: { a: { color: 'green' } },
            a: { color: 'blue' },
            p: { color: 'black' },
          },
          React.createElement(Spy),
        ),
      ),
    );
  });
  it('renders input with style', () => {
    const glitz = new GlitzServer();

    mount(
      React.createElement(
        GlitzProvider,
        { glitz },
        React.createElement(
          StyleProvider,
          {
            universal: { boxSizing: 'border-box' },
            include: { h1: { lineHeight: 1.2 } },
            h1: { margin: { top: 0, bottom: '10px' } },
          },
          React.createElement(styled.H1),
        ),
      ),
    );

    expect(glitz.getStyleMarkup()).toMatchSnapshot();
  });
  it('renders style', () => {
    const Spy: React.FunctionComponent = () => {
      const providedStyle = React.useContext(StyleContext);
      expect(providedStyle).toEqual({
        universal: [{ color: 'red' }],
        a: [{ color: 'green' }, { color: 'blue' }],
        p: [{ color: 'black' }],
      });
      return null;
    };
    mount(
      React.createElement(
        GlitzProvider,
        { glitz: new GlitzClient() },
        React.createElement(
          StyleProvider,
          {
            universal: { color: 'red' },
            include: { a: { color: 'green' } },
            a: { color: 'blue' },
            p: { color: 'black' },
          },
          React.createElement(Spy),
        ),
      ),
    );
  });
  it('concat styles', () => {
    const Spy: React.FunctionComponent = () => {
      const providedStyle = React.useContext(StyleContext);
      expect(providedStyle).toEqual({
        a: [{ color: 'red' }, { color: 'blue' }],
        p: [{ color: 'green' }, { color: 'black' }],
      });
      return null;
    };
    mount(
      React.createElement(
        GlitzProvider,
        { glitz: new GlitzClient() },
        React.createElement(
          StyleProvider,
          { include: { a: { color: 'red' } }, p: { color: 'green' } },
          React.createElement(
            StyleProvider,
            { include: { a: { color: 'blue' } }, p: { color: 'black' } },
            React.createElement(Spy),
          ),
        ),
      ),
    );
  });
});
