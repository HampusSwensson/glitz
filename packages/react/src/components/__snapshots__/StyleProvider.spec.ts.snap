// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react provider renders input with style 1`] = `"<style data-glitz>.a{box-sizing:border-box}.b{line-height:1.2}.c{margin-top:0}.d{margin-bottom:10px}</style>"`;
